File: flutter_app/lib/src/domain/diary.dart
import 'note.dart';
import 'planner.dart';

class DiaryEntry {
  const DiaryEntry({required this.note, required this.date});
  final Note note;
  final DateTime date;
}

class DiaryIndex {
  const DiaryIndex({
    required this.entries,
    required this.days,
    required this.tasks,
  });

  final List<DiaryEntry> entries;
  final Map<String, List<DiaryEntry>> days;
  final Map<String, List<Note>> tasks;
}

abstract final class Diary {
  static const prefix = 'diario_';

  static DateTime validDate(String value) {
    final parsed = parseDate(value);
    if (parsed == null || parsed.year < 1900 || parsed.year > 2200) {
      throw const FormatException('Usa una data AAAA-MM-GG tra 1900 e 2200.');
    }
    return parsed;
  }

  static DateTime? tagDate(String tag) {
    if (!tag.startsWith(prefix)) return null;
    try {
      return validDate(tag.substring(prefix.length));
    } catch (_) {
      return null;
    }
  }

  static DateTime? date(List<String> tags) {
    final dates = tags.map(tagDate).whereType<DateTime>().map(dateKey).toSet();
    if (dates.length != 1) return null;
    return parseDate(dates.single);
  }

  static List<String> datedTags(List<String> tags, DateTime? date) {
    final user = tags.where((tag) => tagDate(tag) == null);
    final next = <String>{...user.map((e) => e.trim()).where((e) => e.isNotEmpty)};
    if (date != null) {
      validDate(dateKey(date));
      next.add('$prefix${dateKey(date)}');
    }
    return next.toList(growable: false);
  }

  static List<String> userTags(List<String> tags) =>
      tags.where((tag) => tagDate(tag) == null).toList(growable: false);

  static DateTime? noteDate(Note note) {
    if (note.isDeleted || note.isTask || note.isVisual) return null;
    return date(note.tags);
  }

  static List<DateTime?> monthCells(DateTime month) {
    final first = DateTime(month.year, month.month, 1);
    final offset = first.weekday - 1;
    final days = DateTime(month.year, month.month + 1, 0).day;
    final size = ((offset + days + 6) ~/ 7) * 7;
    return List.generate(size, (index) {
      final day = index - offset + 1;
      return day >= 1 && day <= days ? DateTime(month.year, month.month, day) : null;
    });
  }

  static DiaryIndex index(List<Note> notes, {String? collectionId}) {
    final entries = <DiaryEntry>[];
    final days = <String, List<DiaryEntry>>{};
    final tasks = <String, List<Note>>{};

    for (final note in notes) {
      final date = noteDate(note);
      if (date != null &&
          (collectionId == null || note.collectionId == collectionId)) {
        final entry = DiaryEntry(note: note, date: date);
        entries.add(entry);
        days.putIfAbsent(dateKey(date), () => <DiaryEntry>[]).add(entry);
      }

      final task = TaskDetails.tryDecode(note.taskJson);
      if (!note.isDeleted && !note.archived && task?.due != null) {
        tasks.putIfAbsent(task!.due!, () => <Note>[]).add(note);
      }
    }

    entries.sort((a, b) {
      final day = b.date.compareTo(a.date);
      if (day != 0) return day;
      final created = b.note.createdAt.compareTo(a.note.createdAt);
      if (created != 0) return created;
      return a.note.id.compareTo(b.note.id);
    });

    for (final list in days.values) {
      list.sort((a, b) {
        final created = b.note.createdAt.compareTo(a.note.createdAt);
        if (created != 0) return created;
        return a.note.id.compareTo(b.note.id);
      });
    }

    return DiaryIndex(entries: entries, days: days, tasks: tasks);
  }
}

File: app/src/main/java/it/notes/ecosystem/domain/Diary.kt
package it.notes.ecosystem.domain

import java.time.LocalDate
import java.time.YearMonth

/** A reserved ordinary tag transports the calendar date through existing backups and sync.
 * The calendar date is deliberately independent of creation/update timestamps and time zones. */
object Diary {
    private const val PREFIX = "diario_"
    fun validDate(value: String): LocalDate {
        require(Regex("[0-9]{4}-[0-9]{2}-[0-9]{2}").matches(value)) { "Usa una data AAAA-MM-GG." }
        return LocalDate.parse(value).also { require(it.year in 1900..2200) { "Scegli un anno tra 1900 e 2200." } }
    }
    fun tagDate(tag: String): LocalDate? = if (!tag.startsWith(PREFIX)) null else runCatching {validDate(tag.removePrefix(PREFIX))}.getOrNull()
    fun date(tags: List<String>): LocalDate? = tags.mapNotNull(::tagDate).distinct().singleOrNull()
    fun datedTags(tags: List<String>, date: LocalDate?): List<String> {
        date?.let {validDate(it.toString())}
        return Tags.normalize(tags.filter {tagDate(it)==null} + listOfNotNull(date?.let {PREFIX+it}))
    }
    fun userTags(tags: List<String>) = tags.filter {tagDate(it)==null}
    fun noteDate(note: Note): LocalDate? = if(note.deletedAt!=null || note.task!=null || note.sketch!=null || PersonalTemplates.TAG in note.tags) null else date(note.tags)
    fun monthCells(month: YearMonth): List<LocalDate?> {
        require(month.year in 1900..2200)
        val offset=month.atDay(1).dayOfWeek.value-1
        val size=((offset+month.lengthOfMonth()+6)/7)*7
        return List(size) {index->(index-offset+1).takeIf {it in 1..month.lengthOfMonth()}?.let(month::atDay)}
    }
    fun index(notes: List<Note>, bookId: String? = null): DiaryIndex {
        val entries=notes.mapNotNull {note->
            val date=noteDate(note) ?: return@mapNotNull null
            if(bookId!=null && note.collectionId!=bookId)return@mapNotNull null
            DiaryEntry(note,date,Attachments.refs(note.body).filter {it.type in listOf(AttachmentType.JPEG,AttachmentType.PNG,AttachmentType.WEBP)})
        }.sortedWith(compareByDescending<DiaryEntry>{it.date}.thenByDescending {it.note.createdAt}.thenBy {it.note.id})
        val tasks=notes.filter {it.deletedAt==null && !it.archived && it.task?.due!=null}.groupBy {LocalDate.parse(it.task!!.due)}
        return DiaryIndex(entries,entries.groupBy {it.date},tasks)
    }
}
data class DiaryEntry(val note: Note,val date: LocalDate,val photos: List<AttachmentRef>)
data class DiaryIndex(val entries: List<DiaryEntry>,val days: Map<LocalDate,List<DiaryEntry>>,val tasks: Map<LocalDate,List<Note>>)
